### ISSUE TITLE
no-undef does not catch when variable is used before it is defined

### ISSUE THREAD

**Issue by lencioni** [eslint/eslint#5437]
**What version of ESLint are you using?**
2.2.0

**What configuration and parser (Espree, Babel-ESLint, etc.) are you using?**
```js
module.exports = {
  rules: {
    'no-undef': 2
  }
};
```

**What did you do? Please include the actual source code causing the issue.**
```js
foo.bar();
var foo = 'foo';
```

```sh
eslint .
```

**What did you expect to happen?**
I expected the eslint output to include:

```
1:1  error  'foo' is not defined  no-undef
```

**What actually happened? Please include the actual, raw output from ESLint.**
There was no eslint output and the process terminated with an exit code of 0.

**Comment by platinumazure** [comment]
The `var` declaration is hoisted above the function call. In addition, `no-undef` is not the rule you want. `no-undef` looks for variables that are not declared anywhere (i.e., global variables) and reports on those; the rule you want is `no-use-before-define`.

Please feel free to write to our mailing list or visit us in chat if you have questions about how to get `no-use-before-define` to work for you.

**Comment by lencioni** [comment]
Ah, that makes sense. I forgot about this hoisting. Thanks!

**Comment by platinumazure** [comment]
No problem. Like I said, `no-use-before-define` will flag your example use case if you want to enforce that, even with hoisting, so do give it a look. :smile:
